core/src/utils/overlays.ts

```diff
@@ -515,7 +515,14 @@ export const dismiss = async (
   }
 
   overlay.el.remove();
-  return true;
+
+  /**
+   * Waits 1 frame after the element has been removed from the DOM to resolve.
+   * Otherwise getTop() may return the dismissed overlay as the topmost overlay.
+   */
+  return new Promise((resolve) => {
```

Comment:
Can you clarify why this is needed? When the overlay is dismissed, we add the `overlay-hidden` class to the host element: https://github.com/ionic-team/ionic-framework/pull/25718/files#diff-e6fe3d2c2375aaff02644f8767f36a3c49825d333f9764644871b0541a95b296R511

The `getOverlay` function, which is called by `getTop` filters out hidden overlays: https://github.com/ionic-team/ionic-framework/blob/f9ab9b54ddb5a3004673e4aaa9cb62fd8e97ba07/core/src/utils/overlays.ts#L350

It does this by excluding overlays that have the `overlay-hidden` class: https://github.com/ionic-team/ionic-framework/blob/f9ab9b54ddb5a3004673e4aaa9cb62fd8e97ba07/core/src/utils/overlays.ts#L105

If we add the `overlay-hidden` class synchronously, why does `getTop` still return a visible overlay?

Reply:
The issue behind this change is triggered by trying to present and dismiss overlays during a SPA navigation event. On navigation start the loading overlay is displayed and on navigation end it is dismissed. For some reason only in Angular, dismissing an overlay in the navigation end recursively through `getTop()`, resolves the dismissed overlay (the classlist has not been mutated at the point in time it is resolved). 

Resolving dismiss into the next frame, prevents this side effect. 

Most of the implementation for the reproduction is here: https://github.com/sean-perkins/gh-25447-overlay-get-top/blob/main/src/app/navigation.service.ts

If we decided against this change, the implementer would need to `requestAnimationFrame` the dismiss method to have the proper overlay resolve from `getTop()`.

Reply:
(Apologies in advance for the essay 😅)

Thanks for the additional context, that was really helpful. I tested using the dev build, but I can still reproduce the infinite loop issue. I used your sample app to test.

------

I took a closer look at why this issue is happening and it looks like `closeLoadingIndicators` is getting called while another `closeLoadingIndicators` operation is in progress. So when navigation ends on `/app/child-2` `closeLoadingIndicators` gets called which begins to dismiss the open loading overlay. During this time `overlay.presented` is set to false:

https://github.com/ionic-team/ionic-framework/blob/abb56d22b4a81d1bc34c689de4ef7218e7503b20/core/src/utils/overlays.ts#L484

However, when the nested router outlet's `ngOnInit` fires, it redirects the app to `/app/child-1`. When the navigation ends on `/app/child-1` `closeLoadingIndicators` gets called again which begins to dismiss the same loading overlay.

Since `overlay.presented` is already `false`, this function returns early: https://github.com/ionic-team/ionic-framework/blob/abb56d22b4a81d1bc34c689de4ef7218e7503b20/core/src/utils/overlays.ts#L478

As a result, `closeLoadingIndicators` is called again and the problem repeats until the loop limit is reached.

I think this is why the `requestAnimationFrame` implemented in the dev's app works. It delays the next `dismiss` call just enough until the first `dismiss` call can fully resolve.

------

If you add logs immediately before and after `loadingIndicator.dismiss()` in `navigation.service.ts` you should see that the first dismiss operation does not resolve until the loop limit is reached. You should see logs that look something like this:

```
begin dismiss <-- first dismiss operation start
begin dismiss
end dismiss
begin dismiss
end dismiss

...

begin dismiss
end dismiss
end dismiss <-- first dismiss operation end
```

-----

At what point is an overlay considered no longer active? (When the dismiss operation begins or when it ends) We set `overlay.presented = false` right when the dismiss operation begins which makes me think that an overlay should no longer be considered active at that point too.

I wonder if we could add a `dismissing` class to the overlay or similar thing right before the async overlay animation is played:

https://github.com/ionic-team/ionic-framework/blob/abb56d22b4a81d1bc34c689de4ef7218e7503b20/core/src/utils/overlays.ts#L499

From there, we could exclude overlays with `.dismissing` from the `getTop` method. Thoughts?